util/src/sync/init_once.rs

```diff
@@ -0,0 +1,296 @@
+            Ordering::Acquire,
+        ) {
+            unreachable!(
+                "InitOnce<{}>: state changed while locked. This is a bug! (state={})",
+                any::type_name::<T>(),
+                actual
+            );
+        }
+        Ok(())
+    }
+
+    /// Initialize the cell to `value`, panicking if it has already been
+    /// initialized.
+    #[track_caller]
+    pub fn init(&self, value: T) {
+        self.try_init(value).unwrap()
+    }
+
+    /// Borrow the contents of this `InitOnce` cell, if it has been
+    /// initialized. Otherwise, if the cell has not yet been initialized, this
+    /// returns `None`.
+    #[inline]
+    pub fn get(&self) -> Option<&T> {
+        if self.state.load(Ordering::Acquire) != INITIALIZED {
+            return None;
+        }
+        unsafe {
+            // Safety: we just checked if the value was initialized.
+            Some(&*((*self.value.get()).as_ptr()))
+        }
+    }
+
+    /// Borrow the contents of this `InitOnce` cell, **without** checking
+    /// whether it has been initialized.
+    ///
+    /// # Safety
+    ///
+    /// The caller is responsible for ensuring that the value has already been
+    /// initialized.
+    ///
+    /// In debug mode, this still checks the state of the cell, so if it has yet
+    /// to be initialized, this will panic. However, in release mode builds,
+    /// this is completely unchecked. If the value has not yet been initialized,
+    /// this may return a pointer to uninitialized memory! It may also return a
+    /// pointer to memory that is currently being written to.
+    ///
+    /// If you see this method panic in debug mode, please, **please** re-check
+    /// your code.
+    #[cfg_attr(not(debug_assertions), inline(always))]
+    #[cfg_attr(debug_assertions, track_caller)]
+    pub unsafe fn get_unchecked(&self) -> &T {
+        debug_assert_eq!(
+            INITIALIZED,
+            self.state.load(Ordering::Acquire),
+            "InitOnce<{}>: accessed before initialized!\n\
+            /!\\ EXTREMELY SERIOUS WARNING: /!\\ This is REAL BAD! If you were \
+            running in release mode, you would have just read uninitialized \
+            memory! That's bad news indeed, buddy. Double- or triple-check \
+            your assumptions, or consider Just Using A Goddamn Mutex --- it's \
+            much safer that way. Maybe this whole `InitOnce` thing was a \
+            mistake...
+            ",
+            any::type_name::<T>(),
+        );
+        unsafe {
+            // Safety: hahaha wheeee no rules! You can't stop meeeeee!
+            &*((*self.value.get()).as_ptr())
+        }
+    }
+}
+
+impl<T> core::ops::Deref for InitOnce<T> {
+    type Target = T;
+
+    #[track_caller]
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        match self.get() {
+            Some(t) => t,
+            None => panic!("InitOnce<{}> not yet initialized!", any::type_name::<T>(),),
+        }
+    }
+}
```

Comment:
I don't love using a panicking `Deref` impl, as calls to `Deref` can be inserted without writing anything in the source text. Perhaps we could have `.get()` with the panicking behaviour, and `.try_get()` which returns an `Option` (or perhaps `unwrap`/`get`?)

Reply:
I like _having_ a `Deref`, though...it's much more of a pain to use these without one. But I see your point...

Reply:
I like *having* a deref impl, though...otherwise, it makes these types much more of a PITA to use.  But, you're probably right that it's not a great idea.